presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPruneIndexSourceColumns.java

```diff
@@ -0,0 +1,113 @@
+import static io.airlift.testing.Closeables.closeAllRuntimeException;
+
+public class TestPruneIndexSourceColumns
+{
+    private RuleTester tester;
+
+    @BeforeClass
+    public void setUp()
+    {
+        tester = new RuleTester();
+    }
+
+    @AfterClass(alwaysRun = true)
+    public void tearDown()
+    {
+        closeAllRuntimeException(tester);
+        tester = null;
+    }
+
+    private PlanNode buildProjectedIndexSource(PlanBuilder p, Predicate<Symbol> projectionFilter)
+    {
+        Symbol orderkey = p.symbol("orderkey", INTEGER);
+        Symbol custkey = p.symbol("custkey", INTEGER);
+        Symbol totalprice = p.symbol("totalprice", DOUBLE);
+        ColumnHandle orderkeyHandle = new TpchColumnHandle(orderkey.getName(), INTEGER);
+        ColumnHandle custkeyHandle = new TpchColumnHandle(custkey.getName(), INTEGER);
+        ColumnHandle totalpriceHandle = new TpchColumnHandle(totalprice.getName(), DOUBLE);
+        return p.project(
+                Assignments.identity(
+                        ImmutableList.of(orderkey, custkey, totalprice).stream()
+                                .filter(projectionFilter)
+                                .collect(toImmutableList())),
+                p.indexSource(
+                        new TableHandle(
+                                new ConnectorId("local"),
+                                new TpchTableHandle("local", "orders", TINY_SCALE_FACTOR)),
+                        ImmutableSet.of(orderkey, custkey),
+                        ImmutableList.of(orderkey, custkey, totalprice),
+                        ImmutableMap.of(
+                                orderkey, orderkeyHandle,
+                                custkey, custkeyHandle,
+                                totalprice, totalpriceHandle),
+                        TupleDomain.fromFixedValues(ImmutableMap.of(totalpriceHandle, asNull(DOUBLE)))));
+    }
+
+    @Test
+    public void testNotAllOutputsReferenced()
+    {
+        tester.assertThat(new PruneIndexSourceColumns())
+                .on(p -> buildProjectedIndexSource(p, symbol -> symbol.getName().equals("orderkey")))
+                .matches(
+                        strictProject(
+                                ImmutableMap.of("x", PlanMatchPattern.expression("orderkey")),
+                                constrainedIndexSource(
+                                        "orders",
+                                        ImmutableMap.of("totalprice", Domain.onlyNull(DOUBLE)),
+                                        ImmutableMap.of(
+                                                "orderkey", "orderkey",
+                                                "totalprice", "totalprice"))));
+    }
+
+    @Test
+    public void testAllOutputsReferenced()
+    {
+        tester.assertThat(new PruneIndexSourceColumns())
+                .on(p -> buildProjectedIndexSource(p, symbol -> true))
+                .doesNotFire();
+    }
+}
```

Comment:
In that case, the rewrite drops the lookup key, and fails an assertion in the IndexSourceNode constructor.  I copied the logic faithfully from PruneUnreferencedOutputs -- it just didn't have any testing, so of course it didn't test this.